// laco/admin/media/views/modal.marko.js
// Compiled using marko@4.2.8 - DO NOT EDIT
"use strict";

var marko_template = module.exports = require("marko/html").t(__filename),
    marko_helpers = require("marko/runtime/html/helpers"),
    marko_loadTag = marko_helpers.t,
    lasso_page_tag = marko_loadTag(require("lasso/taglib/config-tag")),
    lasso_head_tag = marko_loadTag(require("lasso/taglib/head-tag")),
    lasso_body_tag = marko_loadTag(require("lasso/taglib/body-tag")),
    init_components_tag = marko_loadTag(require("marko/components/taglib/init-components-tag")),
    await_reorderer_tag = marko_loadTag(require("marko/taglibs/async/await-reorderer-tag")),
    marko_escapeXmlAttr = marko_helpers.xa;

function render(input, out) {
  var data = input;

  let common = data._common;

  lasso_page_tag({
      dependencies: common.dependencies,
      flags: common.pageFlags,
      dirname: __dirname,
      filename: __filename
    }, out);

  out.w("<!DOCTYPE html><html><head><meta charset=\"UTF-8\"><meta content=\"width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no\" name=\"viewport\"><title>Laco media browser</title>");

  lasso_head_tag({}, out);

  out.w("</head><body class=\"" +
    marko_escapeXmlAttr(common.pageFlags.join(" ")) +
    " theme-cyan\"><section><aside id=\"leftsidebar\" class=\"sidebar media-left-bar\"> <div class=\"menu\"><ul class=\"nav-left-menu list\"><li><a href=\"/media/upload\" class=\"no-pipe\"><i class=\"zmdi zmdi-cloud-upload\"></i><span>Upload</span></a></li><li><a href=\"/logout\" class=\"no-pipe\"><i class=\"zmdi zmdi-info\"></i><span>Help</span></a></li></ul></div></aside><aside id=\"rightsidebar\" class=\"right-sidebar media-right-bar\"><ul class=\"nav nav-tabs tab-nav-right\" role=\"tablist\"><li role=\"presentation\" class=\"active\"><a href=\"#media-actions\" data-toggle=\"tab\">Actions</a></li><li role=\"presentation\"><a href=\"#media-settings\" data-toggle=\"tab\">SETTINGS</a></li></ul><div class=\"tab-content\"><div role=\"tabpanel\" class=\"tab-pane fade in active in active\" id=\"media-actions\"></div><div role=\"tabpanel\" class=\"tab-pane fade\" id=\"media-settings\"><div class=\"demo-settings\"><p>Media SETTINGS</p><ul class=\"setting-list\"><li><span>Report Panel Usage</span><div class=\"switch\"><label><input type=\"checkbox\" checked><span class=\"lever\"></span></label></div></li><li><span>Email Redirect</span><div class=\"switch\"><label><input type=\"checkbox\"><span class=\"lever\"></span></label></div></li></ul></div></div></div></aside></section><section class=\"content\"><div class=\"container-fluid\" style=\"padding: 15px\"><div class=\"row clearfix\"><div id=\"laco-main-content\" class=\"col-lg-12 col-md-12 col-sm-12 col-xs-12\">media modal</div></div></div></section>");

  lasso_body_tag({}, out);

  init_components_tag({}, out);

  await_reorderer_tag({}, out);

  out.w("</body></html>");
}

marko_template._ = render;

marko_template.meta = {
    tags: [
      "lasso/taglib/config-tag",
      "lasso/taglib/head-tag",
      "lasso/taglib/body-tag",
      "marko/components/taglib/init-components-tag",
      "marko/taglibs/async/await-reorderer-tag"
    ]
  };
