// laco/admin/media/views/modal.marko.js
// Compiled using marko@4.2.8 - DO NOT EDIT
"use strict";

var marko_template = module.exports = require("marko/html").t(__filename),
    marko_helpers = require("marko/runtime/html/helpers"),
    marko_loadTag = marko_helpers.t,
    lasso_page_tag = marko_loadTag(require("lasso/taglib/config-tag")),
    marko_str = marko_helpers.s,
    lasso_head_tag = marko_loadTag(require("lasso/taglib/head-tag")),
    lasso_body_tag = marko_loadTag(require("lasso/taglib/body-tag")),
    init_components_tag = marko_loadTag(require("marko/components/taglib/init-components-tag")),
    await_reorderer_tag = marko_loadTag(require("marko/taglibs/async/await-reorderer-tag")),
    marko_escapeXmlAttr = marko_helpers.xa;

function render(input, out) {
  var data = input;

  let common = data._common;

  lasso_page_tag({
      dependencies: common.dependencies,
      flags: common.pageFlags,
      dirname: __dirname,
      filename: __filename
    }, out);

  out.w("<!DOCTYPE html><html><head><meta charset=\"UTF-8\"><meta content=\"width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no\" name=\"viewport\">" +
    marko_str(common.csrf_meta) +
    "<title>Laco media browser</title>");

  lasso_head_tag({}, out);

  out.w("</head><body class=\"" +
    marko_escapeXmlAttr(common.pageFlags.join(" ")) +
    " theme-cyan\"><section class=\"content row clearfix media-modal-content\"><div class=\"col-lg-10 col-md-9 col-sm-6 col-xs-6 media-modal-main\"><ul class=\"nav nav-tabs tab-nav-right media-main-tab\" role=\"tablist\"><li role=\"presentation\" class=\"active\"><a href=\"#media-browse\" data-toggle=\"tab\">Browse Medias</a></li><li role=\"presentation\"><a href=\"#media-upload\" data-toggle=\"tab\">Upload</a></li></ul><div class=\"tab-content\"><div role=\"tabpanel\" class=\"tab-pane fade in active\" id=\"media-browse\"></div><div role=\"tabpanel\" class=\"tab-pane fade row clearfix laco-upload-holder\" id=\"media-upload\" style=\"height: Calc(100% - 41px)\"><div class=\"dz-message tips\"><h3>Drag and drop medias here</h3><div>or</div><button type=\"button\" class=\"btn btn-primary m-t-10 m-b-30 waves-effect\">Browse media files </button></div></div></div></div><aside id=\"rightsidebar\" class=\"col-lg-2 col-md-3 col-sm-6 col-xs-6 media-right-bar\"></aside><div id=\"laco-dz-preview-tpl\"><div class=\"col-lg-2 col-md-3 col-sm-6 col-xs-12 dz-preview dz-file-preview media-thumb\"><div class=\"dz-details detail\"><div class=\"thumb\"><img data-dz-thumbnail></div><div class=\"dz-progress progress\"><span class=\"dz-upload progress-bar progress-bar-striped active\" role=\"progressbar\" aria-valuenow=\"0\" aria-valuemin=\"0\" aria-valuemax=\"100\" data-dz-uploadprogress></span></div><div class=\"info\"><div class=\"dz-filename name\"><span data-dz-name></span></div><div class=\"dz-size size\" data-dz-size></div></div></div><div class=\"remove\"></div><div class=\"dz-mark dz-success-mark\"></div><div class=\"dz-mark dz-error-mark\"></div><div class=\"dz-error-message message\"><span data-dz-errormessage></span></div></div></div></section>");

  lasso_body_tag({}, out);

  init_components_tag({}, out);

  await_reorderer_tag({}, out);

  out.w("</body></html>");
}

marko_template._ = render;

marko_template.meta = {
    tags: [
      "lasso/taglib/config-tag",
      "lasso/taglib/head-tag",
      "lasso/taglib/body-tag",
      "marko/components/taglib/init-components-tag",
      "marko/taglibs/async/await-reorderer-tag"
    ]
  };
