// laco/index.js
"use strict";
global.LACO_ROOT = __dirname;
const path = require('path');
const dotenv = require('dotenv');
const nconf = require('nconf');
const requirement = require('./libs/requirement');
const logger = require('./libs/logger');
const helper = require('./libs/helper');
const core = require('./libs/core');
const packer = lacoLib('./libs/packer');
const cli = require('./libs/cli');

let init = appConfigFile => {
    cli.init();
    if (cli.isCommand('config')) return packer.config(err => logger.info('Laco configuration completed!'));

    logger.info('Laco is initializing...');
    dotenv.config({path: process.env.APP_ROOT + '/.env'});
    nconf.argv().env().file({ file: appConfigFile });

    logger.info('Laco is checking for requirements...');
    requirement.check({
        port: nconf.get('APP_PORT'),
        storage: nconf.get('STORAGE'),
        cache: nconf.get('CACHE')
    }, (err) => {
        if(err) return logger.error('Laco cannot start because an error occured: ' + err);
        requirement.dbInit({
            uri: nconf.get('DB_CONNECT_URI'),
            options: {}
        });
        core.boot();
    });
}

exports.init = init;
