// laco/index.js
'use strict';

const _ = require('lodash');
const http = require('http');
const dotenv = require('dotenv');
const express = require('express');
const socketIO = require('socket.io');

class Laco {
    constructor(options = {}) {
        this.config(options);

        this.app = this.options.app || express();
        this.server = http.Server(this.app);
        this.socket = null;

        if (this.options.socket.enabled) {
            this.socket = socketIO(this.server, this.options.socket);
        }

        this.app.get('/', (req, res) => {
            res.end('ok');
        });
    }

    config(options = {}) {
        this.extensions = {
            loaded: [],
            enabled: [],
            disabled: []
        };
        this.options = _.merge({
            app: null,
            listen: {
                ip: '127.0.0.1',
                port: 3000,
                secure: false,
            },
            socket: {
                enabled: true,
                path: '/socket.io',
                serveClient: true,
                origins: '*'
            },
            extension: {
                autoLoad: true,
                autoUpdate: false,
                actions: {
                    route: ['read', 'register', 'unregister'],
                    midlleware: ['read', 'register', 'unregister'],
                    view: ['read', 'register', 'unregister'],
                    theme: ['read', 'register', 'unregister']
                }
            }
        }, options);
    }

    start() {
        this.server.listen(
            this.options.listen.port,
            this.options.listen.ip,
            () => {
                process.stdout.write(`App running on port: ${this.options.listen.port}`);
            }
        );
    }
}

module.exports = Laco;
