// laco/index.js
'use strict';

const _ = require('lodash');
const http = require('http');
const dotenv = require('dotenv');
const express = require('express');
const socketIO = require('socket.io');
const logger = require('./libs/logger');
const router = require('./libs/router');
const subapp = require('./libs/subapp');
const helper = require('./libs/helper');
const ExtManager = require('./libs/extension/manager');
const EventEmitter = require('./libs/EventEmitter');

class Laco extends EventEmitter {
    constructor(options = {}) {
        super();
        this.config(options);

        this.app = this.options.app || express();
        this.server = http.Server(this.app);
        this.socket = null;
        this.logger = logger;
        this.router = router;
        this.SubApp = subapp.Base;
        this.subAppManager = new subapp.SubAppManager(this.app);

        if (this.options.socket.enabled) {
            this.socket = socketIO(this.server, this.options.socket);
        }

        this.extManager = new ExtManager(this, this.options.extension);
        this.all(['ext.manager.initialized'], () => this.emit('ready'));
    }

    config(options = {}) {
        this.extensions = {
            loaded: [],
            enabled: [],
            disabled: []
        };
        this.options = _.merge({
            app: null,
            listen: {
                ip: '127.0.0.1',
                port: 3000,
                secure: false,
            },
            socket: {
                enabled: true,
                path: '/socket.io',
                serveClient: true,
                origins: '*'
            },
            extension: {
                autoLoad: true,
                autoUpdate: false,
                actions: {
                    route: ['read', 'register', 'unregister'],
                    midlleware: ['read', 'register', 'unregister'],
                    view: ['read', 'register', 'unregister'],
                    theme: ['read', 'register', 'unregister']
                }
            }
        }, options);
    }

    start() {
        this.subAppManager.dispatch(this.app);
        this.server.listen(
            this.options.listen.port,
            this.options.listen.ip,
            () => {
                logger.info(`[start] App running on port: ${this.options.listen.port}`);
                this.emit('started', this);
            }
        );
    }
}

exports.Laco = Laco;
exports.SubApp = subapp.Base;
