// laco-ext-router/views/login.marko.js
// Compiled using marko@4.2.8 - DO NOT EDIT
"use strict";

var marko_template = module.exports = require("marko/html").t(__filename),
    marko_helpers = require("marko/runtime/html/helpers"),
    marko_loadTag = marko_helpers.t,
    lasso_page_tag = marko_loadTag(require("../../laco/node_modules/lasso/taglib/config-tag")),
    lasso_head_tag = marko_loadTag(require("../../laco/node_modules/lasso/taglib/head-tag")),
    marko_str = marko_helpers.s,
    lasso_body_tag = marko_loadTag(require("../../laco/node_modules/lasso/taglib/body-tag")),
    init_components_tag = marko_loadTag(require("marko/components/taglib/init-components-tag")),
    await_reorderer_tag = marko_loadTag(require("marko/taglibs/async/await-reorderer-tag"));

function render(input, out) {
  var data = input;

  let common = data._common;

  lasso_page_tag({
      dependencies: common.dependencies,
      flags: common.pageFlags,
      dirname: __dirname,
      filename: __filename
    }, out);

  out.w("<!DOCTYPE html><html><head><meta charset=\"UTF-8\"><meta content=\"width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no\" name=\"viewport\"><title>Login - Laco</title>");

  lasso_head_tag({}, out);

  out.w("</head><body class=\"login-page\"><div class=\"login-box\"><div class=\"logo\"><a href=\"javascript:void(0);\"><b>Laco</b> Admin</a><small>Fast - Ligh Weight Node.js CMS</small></div><div class=\"card\"><div class=\"body\"><form id=\"sign_in\" method=\"POST\">" +
    marko_str(data._common.csrf_field) +
    "<div class=\"msg\">Sign in to start your session</div><div class=\"input-group\"><span class=\"input-group-addon\"><i class=\"zmdi zmdi-account-circle\"></i></span><div class=\"form-line\"><input type=\"text\" class=\"form-control\" name=\"username\" placeholder=\"Username\" required autofocus></div></div><div class=\"input-group\"><span class=\"input-group-addon\"><i class=\"zmdi zmdi-lock\"></i></span><div class=\"form-line\"><input type=\"password\" class=\"form-control\" name=\"password\" placeholder=\"Password\" required></div></div><div class=\"row\"><div class=\"col-xs-8 p-t-5\"><input type=\"checkbox\" name=\"rememberme\" id=\"rememberme\" class=\"filled-in chk-col-cyan\"><label for=\"rememberme\">Remember Me</label></div><div class=\"col-xs-4\"><button class=\"btn btn-block bg-cyan waves-effect\" type=\"submit\">SIGN IN</button></div></div><div class=\"row m-t-15 m-b--20\"><div class=\"col-xs-6\"><a href=\"sign-up.html\">Register Now!</a></div><div class=\"col-xs-6 align-right\"><a href=\"forgot-password.html\">Forgot Password?</a></div></div></form></div></div></div>");

  lasso_body_tag({}, out);

  init_components_tag({}, out);

  await_reorderer_tag({}, out);

  out.w("</body></html>");
}

marko_template._ = render;

marko_template.meta = {
    tags: [
      "../../laco/node_modules/lasso/taglib/config-tag",
      "../../laco/node_modules/lasso/taglib/head-tag",
      "../../laco/node_modules/lasso/taglib/body-tag",
      "marko/components/taglib/init-components-tag",
      "marko/taglibs/async/await-reorderer-tag"
    ]
  };
