// laco/libs/post/model.js
"use strict";
const mongoose = require('mongoose');
const path = require('path');
const autoIncrement = require('mongoose-auto-increment');
const g = lacoLib('./libs/global');

let register = () => {
    g.schemas.post = new mongoose.Schema({
        name:  String,
        url:   String,
        image:   String,
        maintax: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'taxonomies',
            required: false
        },
        subtaxes: {
            type: [mongoose.Schema.Types.ObjectId],
            ref: 'taxonomies',
            required: false
        },
        content:   String,
        type: {
            type: 'String',
            default: 'post',
            required: true
        },
        fields: {},
        tags:  {
            type: [String]
        },
        properties: {
            total_views: 0,
            total_comments: 0,
            total_likes: 0,
            total_rates: 0,
            rate_score: 0
        },
        add_date: {
            type: Date,
            default: Date.now
        },
        publish_date: {
            type: Date,
            default: Date.now
        },
        update_date: {
            type: Date,
            default: Date.now
        },
        expired_date: {
            type: Date,
            required: false
        },
        edit_history: [
            {
                time: {
                    type: Date,
                    default: Date.now,
                    required: true
                },
                content: {
                    type: String,
                    required: true
                },
                user: {
                    type: mongoose.Schema.Types.ObjectId,
                    ref: 'User',
                    required: true
                }
            }
        ],
        user: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'User',
            required: true
        }
    });

    g.schemas.post.index({
        url: "text",
        name: "text",
        tags: "text"
    }, {
        weights: {
            url: 10,
            name: 8,
            tags: 5
        },
        name: 'postTextIndex'
    });
    g.schemas.post.plugin(autoIncrement.plugin, {
        model: 'posts',
        field: 'intId',
        startAt: 1,
        incrementBy: 1
    });

    g.models.post = mongoose.model('posts', g.schemas.post);
};

exports.register = register;
