// laco/libs/media/model.js
"use strict";

"use strict";
const mongoose = require('mongoose');
const path = require('path');
const autoIncrement = require('mongoose-auto-increment');
const g = lacoLib('./libs/global');
const MediaObject = lacoLib('./libs/media/object');

let register = () => {
    g.schemas.media = new mongoose.Schema({
        name:  {
            type: String,
            required: true
        },
        user: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'User',
            required: false
        },
        description: String,
        alt: String,
        source: String,
        encoding: String,
        mimetype: String,
        size: Number,
        date: {
            type: Date,
            default: Date.now
        }
    });

    g.schemas.media.post('find', (doc, next) => {
        console.log('find');
        //doc = new MediaObject(doc);
        console.log(doc);
        next();
    });

    g.models.media = mongoose.model('medias', g.schemas.media);
};

exports.register = register;
