// laco/libs/admin/module.js
"use strict";

const path = require('path');
const _ = require('underscore');
const cheerio = require('cheerio')
const nconf = require('nconf');
const helper = lacoLib('./libs/helper');
const Module = lacoLib('./libs/module');
const adminMenu = require('./menu');
const assetDependencies = require('./asset');
let theme = {
    default_layout: null,
    pipe_layout: null
}

class AdminModule extends Module {
    constructor(moduleName) {
        theme.default_layout = lacoLib('./data/views/admin/default/layout_default.marko');
        theme.pipe_layout = lacoLib('./data/views/admin/default/layout_pipe.marko');

        let caller = helper.caller();
        let moduleDir = path.dirname(caller.getFileName());
        super(moduleDir, moduleName || caller.getFunctionName());

        this.dependencies = [];
    }
    registerMenus(menus) {
        adminMenu.add(menus);
    }

    addDepencencies(deps) {
        if(!_isArray(deps)) deps = [deps];
        _.each(deps, dep => this.dependencies.push(_.isString(dep) ? {path: dep, 'if-flag': `module-${this.moduleAlias}`} : dep));
    }

    link(action, parameters) {
        let params = [];
        if(!_.isObject(parameters)) parameters = {};
        let link = `/${nconf.get('APP_ADMIN_DIR')}/${this.moduleAlias}`;
        if(action) link += `/${action}`;
        _.each(parameters, (v, k) => params.push(`${k}=${v}`))
        if(!!params.length) link += '?' + parameters.join('&');
        return link;
    }
    view(viewName, data, req, res) {
        //if not piped request, add common dependencies
        let isPipe = !!req.pipe;
        let viewDependencies = [];
        if(!isPipe) {
            viewDependencies = this.dependencies;
            _.each(assetDependencies.commonDependencies, dep => viewDependencies.push({path: dep, 'if-flag': 'laco-common'}));
        }
        else viewDependencies = [];

        let pipeFlag = isPipe ? 'laco-pipe' : 'laco-common';

        let caller = helper.caller();
        this.moduleAction = caller.getFunctionName();
        data._common = {};
        data._common.module = {
            name: this.moduleName,
            alias: this.moduleAlias,
            action: this.moduleAction,
            view: viewName
        }
        data._common.layout = (isPipe && theme.pipe_layout) ? theme.pipe_layout : theme.default_layout;
        data._common.dependencies = viewDependencies;
        data._common.pageFlags = [
            pipeFlag,
            `module-${this.moduleAlias}`,
            `action-${this.moduleAction}`,
            `view-${viewName}`,
            `${this.moduleAlias}-${this.moduleAction}-${viewName}`
        ];
        let viewObj = super.view(viewName, data, req, res);
        if(isPipe) {
            viewObj
                .then(content => {
                    let allScripts = [];
                    let $ = cheerio.load(content.toString(), {decodeEntities: false});
                    $('script').each((i, elem) => {
                        let scriptSrc = $(elem).attr('src');
                        if(scriptSrc) {
                            allScripts.push(scriptSrc);
                            $(elem).remove();
                        }
                    });

                    let loadedObj = {
                        html: $.html().replace(/\n/g, ''),
                        scripts: allScripts,
                        target: req.originalUrl.replace(`?pipe=${req.pipe}`, '').replace(`&pipe=${req.pipe}`, '')
                    };
                    let html = `<script>var lacoLoadedPipe = ${JSON.stringify(loadedObj)};parent.pipeContentExtractor(lacoLoadedPipe);</script>`;
                    res.end(html);
                })
                .catch(e => {
                    logger.error(`[AdminModule.View.Pipe] Error render pipe view: ${e.message}`);
                    res.end('Error!!!');
                });
        }
    }
}
module.exports = AdminModule;
