// laco/libs/subapp/base.js
'use strict';

const _ = require('lodash');
const path = require('path');
const express = require('express');

const logger = require('../logger');
const helper = require('../helper');

const RouterManager = require('../router');
const ControllerManager = require('../controller');
const UIManager = require('../ui').Manager;

class SubAppBase {
    constructor(laco, name = '', options = {}) {
        this.laco = laco;
        this.caller = helper.caller();
        this.name = name || this.caller.baseDir.trimChar('/').replace('node_modules/laco-ext-', 'ext::');
        this.options = _.merge({
            mvc: {
                enabled: true,
                useUI: true,
                controllerPath: path.join(this.caller.dir, 'controllers')
            }
        }, options);

        this.express = express();
        this.router = new RouterManager(this);
        this.controller = new ControllerManager(this);
        this.ui = this.options.mvc.useUI ? new UIManager(this) : null;
    }

    dispatch() {
        this.router.dispatch();
    }
}

module.exports = SubAppBase;
