// laco/libs/subapp/index.js
'use strict';

const _ = require('lodash');
const Router = require('../router');
const logger = require('../logger');
const helper = require('../helper');
const SubAppBase = require('./base');

class SubAppManager {
    constructor(laco) {
        this.laco = laco;
        this.app = laco.app;
        this.registeredApps = {};
    }

    all() {
        return this.registeredApps;
    }

    get(base) {
        let cleanedBase = base.trimChar('/');
        if (!this.registeredApps[cleanedBase]) {
            return logger.error(`[SubApp.register] App base not found: ${base}`);
        }
        return this.registeredApps[cleanedBase];
    }

    register(base = null, app) {
        if (!base || !_.isString(base)) {
            return logger.error(`[SubApp.register] Invalid app base(tring): (empty)`);
        }
        let cleanedBase = base.trimChar('/');
        if (this.registeredApps[cleanedBase]) {
            return logger.error(`[SubApp.register] App base existed: ${base}`);
        }
        logger.info(`[SubApp.register] [${app.name}] Success register sub app: ${base}`);
        this.registeredApps[cleanedBase] = {
            enabled: true,
            base: base,
            app: app
        };
    }

    delete(base) {
        logger.info(`[SubApp.register] Deleting sub app: ${base}`);
        let cleanedBase = base.trimChar('/');
        if (!this.registeredApps[cleanedBase]) {
            return logger.error(`[SubApp.register] App base not found: ${base}`);
        }
        delete this.registeredApps[cleanedBase];
    }

    dispatch() {
        _.each(this.registeredApps, sub => {
            sub.app.dispatch();
            sub.app.express.on('mount', parent => {
                logger.info(`[SubApp.dispatch] [${sub.base.trimChar('/')}] Mounted...`);
            });
            this.app.use(sub.base, sub.app.express);
        });
    }
}

exports.Base = SubAppBase;
exports.SubAppManager = SubAppManager;
