// laco/libs/taxonomy/default.js
"use strict";

const category = {
    label: 'Category',
    name: 'category',
    properties: {
        name: {
            label: lang('Name'),
            helper: lang('Enter category name'),
            type: 'String',
            required: true,
            display: 'input',
            position: 'topBox',
            cols: 6
        },
        url: {
            label: lang('Url'),
            type: 'String',
            default: '@name|helper.cleanUrl',
            display: 'input',
            position: 'topBox',
            cols: 6
        },
        image: {
            label: lang('Image'),
            type: 'String',
            display: 'image',
            visible: true,
            position: 'imgBox',
            cols: 12
        },
        parent: {
            label: lang('Parent taxonomy'),
            type: 'Referer',
            ref: 'taxonomy',
            ref_type: 'category',
            display: 'referer_single_select',
            position: 'rightSidebar',
            exclude: 'self',
            visible: true,
            cols: 12
        },
        content: {
            label: lang('Content'),
            type: 'String',
            display: 'html',
            position: 'contentBox',
            visible: true,
            cols: 12
        }
    },
    fields: {
        description: {
            label: lang('Description'),
            type: 'String',
            display: 'textarea',
            position: 'topBox',
            visible: true,
            cols: 12
        },
        meta_titlle: {
            label: lang('Meta title'),
            type: 'String',
            display: 'meta_title',
            position: 'topBox',
            visible: true,
            cols: 6
        },
        meta_description: {
            label: lang('Meta description'),
            type: 'String',
            display: 'meta_description',
            position: 'topBox',
            visible: true,
            cols: 6
        }
    }
}

exports.category = category;
