// laco/admin/taxonomy/index.js
"use strict";

const AdminModule = lacoLib('./libs/admin/module');
const taxonomy = lacoLib('./libs/taxonomy');

class Taxonomy extends AdminModule {
    constructor(laco) {
        super(laco);
        this.registerViews([
            'main',
            'add'
        ]);

        this.registerMenus([
            {
                name: 'taxonomy_main',
                anchor: 'Taxonomy',
                link: this.link(),
                icon: '<i class="zmdi zmdi-collection-text"></i>',
                priority: 8
            },
            {
                name: 'taxonomy_add',
                parent: 'taxonomy_main',
                anchor: 'Add taxonomy',
                link: this.link('add'),
                icon: '<i class="zmdi zmdi-plus-circle"></i>'
            }
        ]);
    }

    main(req, res) {
        this.view('main', {}, req, res);
    }

    add(req, res) {
        let taxType = req.query.type || 'category';
        let taxId = req.params.extra ? 1*req.params.extra : null;
        let taxItem = new taxonomy.instant(taxId, taxType);

        if(req.method == 'POST') {
            delete req.body._csrf;
            taxItem.save(req.body)
                .then(tax => res.redirect(taxItem.editUrl))
                .catch(e => this.view('add', {taxItem, taxId, taxType}, req, res));
            return;
        }

        this.view('add', {taxItem, taxId, taxType}, req, res);
    }
}
module.exports = Taxonomy;
