// laco/admin/taxonomy/index.js
"use strict";

const AdminModule = lacoLib('./libs/admin/module');
const taxonomy = lacoLib('./libs/taxonomy');

class Taxonomy extends AdminModule {
    constructor() {
        super();
        this.registerViews([
            'main',
            'add'
        ]);

        this.registerMenus([
            {
                name: 'taxonomy_main',
                anchor: 'Taxonomy',
                link: this.link(),
                icon: '<i class="zmdi zmdi-collection-text"></i>'
            },
            {
                name: 'taxonomy_add',
                parent: 'taxonomy_main',
                anchor: 'Add taxonomy',
                link: this.link('add'),
                icon: '<i class="zmdi zmdi-plus-circle"></i>'
            }
        ]);
    }

    main(req, res) {
        this.view('main', {}, req, res);
    }

    add(req, res) {
        let taxType = req.query.type || 'category';
        let taxId = req.params.extra ? 1*req.params.extra : null;
        let taxItem = new taxonomy.instant(taxId, taxType);
        this.view('add', {taxItem, taxId, taxType}, req, res);
    }
}
module.exports = Taxonomy;
