// laco/libs/admin_module.js
"use strict";

const path = require('path');
const _ = require('underscore');
const nconf = require('nconf');
const Module = lacoLib('./libs/module');
const adminMenu = require('./admin/menu');
let theme = {
    layout: null
}

class AdminModule extends Module {
    constructor(moduleDir, moduleName) {
        theme.layout = lacoLib('./data/views/admin/default/layout_default.marko');
        super(moduleDir, moduleName);
        const adminAsset = path.join(LACO_ROOT, 'data/assets/admin');
        let commonDeps = [
            path.join(adminAsset, 'plugins/material-design-iconic-font/css/material-design-iconic-font.min.css'),
            path.join(adminAsset, 'plugins/bootstrap/css/bootstrap.css'),
            path.join(adminAsset, 'plugins/node-waves/waves.css'),
            path.join(adminAsset, 'plugins/animate-css/animate.css'),
            path.join(adminAsset, 'css/style.css'),
            path.join(adminAsset, 'css/custom.css'),
            path.join(adminAsset, 'css/themes/all-themes.css'),
            path.join(adminAsset, 'plugins/jquery/jquery.min.js'),
            path.join(adminAsset, 'plugins/bootstrap/js/bootstrap.js'),
            path.join(adminAsset, 'plugins/bootstrap-select/js/bootstrap-select.js'),
            path.join(adminAsset, 'plugins/jquery-slimscroll/jquery.slimscroll.js'),
            path.join(adminAsset, 'plugins/node-waves/waves.js'),
            path.join(adminAsset, 'js/admin.js'),
            path.join(adminAsset, 'js/demo.js')
        ];
        this.dependencies = [];
        for(let dep of commonDeps) {
            this.dependencies.push({
                path: dep,
                'if-flag': 'laco-common'
            });
        }
    }
    registerMenus(menus) {
        adminMenu.add(menus);
        adminMenu.toNested();
    }
    link(action, parameters) {
        if(!_.isObject(parameters)) parameters = {};
        let moduleCleanName = path.basename(this.moduleDir);
        let link = `/${nconf.get('APP_ADMIN_DIR')}/${moduleCleanName}`;
        if(action) link += `/${action}`;
        let params = [];
        for(let k in parameters) {
            parameters.push(`${k}=${parameters[k]}`);
        }
        if(!!params.length) link += '?' + parameters.join('&');
        return link;
    }
    view(viewName, data, req, res) {
        data._common = {};
        data._common.layout = theme.layout;
        data._common.dependencies = this.dependencies;
        data._common.pageFlags = ['laco-common'];
        super.view(viewName, data, req, res);
    }
}
module.exports = AdminModule;
