// laco/libs/module.js
"use strict";

const path = require('path');
const async = require('async');
const _ = require('underscore');
const logger = require('./logger');

class Module {
    constructor(moduleDir, moduleName) {
        this.moduleName = moduleName || path.basename(moduleDir);
        this.moduleDir = moduleDir;
        this.moduleAlias = path.basename(moduleDir);
        this.moduleAction = null;
        this.viewsObj = {}
    }
    registerViews(views) {
        if(!_.isArray(views)) {
            logger.error(`[Module.registerViews]--> ${this.moduleName}: Invalid views array: `);
            logger.error(views);
            return;
        }
        async.map(views, (view, callback) => {
            if(!_.isString(view)) {
                logger.error(`[Module.registerViews]--> ${this.moduleName}: Invalid view item: `);
                logger.error(view);
                return callback(null, null);
            }

            let viewFile = path.join(this.moduleDir, `views/${view}.marko`);
            logger.info(`Registering view for extension "${this.moduleName}": ${view}`);
            try {
                this.viewsObj[view] = require(viewFile);
                callback(null);
            }
            catch(e) {
                //logger.error(`[Module.registerViews]--> ${this.moduleName}: View '${view}' not found, with message: ${e.message}`);
                logger.error(`[Module.registerViews]--> ${this.moduleName}: Error registering view '${view}'`);
                callback(null, null);
            }
        }, (err, _views) => {
            if(err) {
                logger.error(`[Module.registerViews]--> ${this.moduleName}: An error occured when registering module views: `);
                logger.error(err);
            }
        });
        
    }
    view(viewName, data, req, res) {
        let isPipe = !!req.pipe;
        return isPipe ? this.viewsObj[viewName].render(data) : this.viewsObj[viewName].render(data, res);
    }
}

module.exports = Module;
