// laco/admin/dashboard/index.js
"use strict";

const AdminModule = lacoLib('./libs/admin/module');

class Dashboard extends AdminModule {
    constructor(laco) {
        super(laco);
        this.registerViews([
            'main',
            'setting'
        ]);

        this.registerMenus([
            {
                name: 'dashboard_main',
                anchor: 'Dashboard',
                link: this.link(),
                icon: '<i class="zmdi zmdi-view-dashboard"></i>',
                priority: 9
            },
            {
                name: 'dashboard_viewsite',
                parent: 'dashboard_main',
                anchor: 'View site',
                link: '/',
                icon: '<i class="zmdi zmdi-arrow-right-top"></i>',
                priority: 9
            },
            {
                name: 'dashboard_setting',
                parent: 'dashboard_main',
                anchor: 'Setting',
                link: this.link('setting'),
                icon: '<i class="zmdi zmdi-settings"></i>'
            }
        ]);
    }

    main(req, res) {
        this.view('main', {}, req, res);
    }
}
module.exports = Dashboard;
