// laco/admin/media/index.js
"use strict";

const AdminModule = lacoLib('./libs/admin/module');

class Media extends AdminModule {
    constructor() {
        super();
        this.registerViews([
            'listing',
            'upload',
            'modal'
        ]);

        this.registerMenus([
            {
                name: 'media_main',
                anchor: 'Media',
                link: this.link(),
                icon: '<i class="zmdi zmdi-storage"></i>',
                priority: 1
            },
            {
                name: 'media_upload',
                parent: 'media_main',
                anchor: 'Upload',
                link: '/',
                icon: '<i class="zmdi zmdi-upload"></i>',
                priority: 1
            },
            {
                name: 'media_setting',
                parent: 'media_main',
                anchor: 'Setting',
                link: this.link('setting'),
                icon: '<i class="zmdi zmdi-settings"></i>'
            }
        ]);
    }

    main(req, res) {
        this.view('main', {}, req, res);
    }
}
module.exports = Media;
