// laco/admin/media/index.js
"use strict";

const path = require('path');
const AdminModule = lacoLib('./libs/admin/module');
const logger = lacoLib('./libs/logger');
const media = lacoLib('./libs/media');
const globalAssetPath = path.join(LACO_ROOT, 'data/assets');
const uploader = require('./uploader');

class Media extends AdminModule {
    constructor(laco) {
        super();

        this.registerViews([
            'listing',
            'upload',
            'modal'
        ]);

        this.registerMenus([
            {
                name: 'media_main',
                anchor: 'Media',
                link: this.link(),
                icon: '<i class="zmdi zmdi-storage"></i>',
                priority: 1
            },
            {
                name: 'media_upload',
                parent: 'media_main',
                anchor: 'Upload',
                link: '/',
                icon: '<i class="zmdi zmdi-upload"></i>',
                priority: 1
            },
            {
                name: 'media_setting',
                parent: 'media_main',
                anchor: 'Setting',
                link: this.link('setting'),
                icon: '<i class="zmdi zmdi-settings"></i>'
            }
        ]);

        this.addDepencencies([
            path.join(globalAssetPath, 'vendors/dropzone-4.3.0/dropzone.min.css'),
            path.join(globalAssetPath, 'vendors/dropzone-4.3.0/basic.min.css'),
            path.join(globalAssetPath, 'vendors/dropzone-4.3.0/dropzone.min.js'),
            path.join(globalAssetPath, 'css/media.scss'),
            path.join(globalAssetPath, 'js/media.js'),
        ]);

        let thumbRouter = () => {
            let router = new laco.express.Router()

            router.get('/data/thumbs/:filename([a-zA-Z0-9\-\_\/\.\%]+)', (req, res) => {
                media.tools.thumbGenerator(req.params.filename)
                    .then(thumbStream => thumbStream.pipe(res))
                    .catch(err => console.log(err));
            });

            return router
        }

        laco.app.use(thumbRouter());
    }

    main(req, res) {
        this.view('listing', {}, req, res);
    }

    modal(req, res) {
        let targetField = req.query.target || null;
        let isMulti = !!(1*req.query.multi) || false;
        let types = req.query.types || 'all';
        let mediasProvider = media.provider.paging({
            page: req.paging.current,
            types: types.split('.')
        });
        this.view('modal', {
            targetField,
            isMulti,
            types,
            mediasProvider
        }, req, res, 'blank');
    }

    upload(req, res) {
        uploader.uploadHandler(req, res, (err, _file) => {
            if(err) {
                logger.error(`[Module.media.upload] Error uploading media: ${req.file}`);
                logger.error(req.file);
                logger.error(err);
            }

            res.json({
                status: err ? false : true,
                message: err ? err.message : 'File uploaded',
                result: _file || {}
            });
        });
    }
}
module.exports = Media;
