// laco/admin/media/index.js
"use strict";

const path = require('path');
const AdminModule = lacoLib('./libs/admin/module');
const globalAssetPath = path.join(LACO_ROOT, 'data/assets');

class Media extends AdminModule {
    constructor() {
        super();

        this.registerViews([
            'listing',
            'upload',
            'modal'
        ]);

        this.registerMenus([
            {
                name: 'media_main',
                anchor: 'Media',
                link: this.link(),
                icon: '<i class="zmdi zmdi-storage"></i>',
                priority: 1
            },
            {
                name: 'media_upload',
                parent: 'media_main',
                anchor: 'Upload',
                link: '/',
                icon: '<i class="zmdi zmdi-upload"></i>',
                priority: 1
            },
            {
                name: 'media_setting',
                parent: 'media_main',
                anchor: 'Setting',
                link: this.link('setting'),
                icon: '<i class="zmdi zmdi-settings"></i>'
            }
        ]);

        this.addDepencencies([
            path.join(globalAssetPath, 'vendors/dropzone-4.3.0/dropzone.min.css'),
            path.join(globalAssetPath, 'vendors/dropzone-4.3.0/basic.min.css'),
            path.join(globalAssetPath, 'vendors/dropzone-4.3.0/dropzone.min.js'),
            path.join(globalAssetPath, 'css/media.scss'),
            path.join(globalAssetPath, 'js/media.js'),
        ]);
    }

    main(req, res) {
        this.view('listing', {}, req, res);
    }

    modal(req, res) {
        this.view('modal', {}, req, res, 'blank');
    }
}
module.exports = Media;
