// laco/libs/packer/config_rules.js
"use strict";

let rules = {
    properties: {
        APP_PORT: {
            description: 'Please enter app port',
            type: 'integer',
            pattern: /^[0-9]+$/,
            message: 'App port must be an integer',
            default: 2811,
            required: true
        },
        APP_ADMIN_DIR: {
            description: 'Please enter admin route dir',
            type: 'string',
            pattern: /^[a-zA-Z0-9\-\_]+$/,
            message: 'Admin dir must be only letters, number, or dashes',
            default: 'admin',
            required: true
        },
        APP_USE_PROXY: {
            description: 'Please select proxy system',
            type: 'string',
            message: 'Supported proxy system: nginx',
            default: 'none',
            required: false
        },
        DB_CONNECT_URI: {
            description: 'Enter mongodb connect uri',
            type: 'string',
            default: 'mongodb://laco_user:123456789@localhost/laco',
            required: true
        },
        DB_OPTIONS_poolSize: {
            description: 'Enter mongodb poolsize',
            type: 'integer',
            default: 10,
            required: true
        }
    }
};

exports.rules = rules;
