// laco/libs/packer/config_rules.js
"use strict";

const helper = lacoLib('./libs/helper');

let rules = {
    properties: {
        APP_PORT: {
            description: 'Please enter app port',
            type: 'integer',
            pattern: /^[0-9]+$/,
            message: 'App port must be an integer',
            default: 2811,
            required: true
        },
        APP_ADMIN_DIR: {
            description: 'Please enter admin route dir',
            type: 'string',
            pattern: /^[a-zA-Z0-9\-\_]+$/,
            message: 'Admin dir must be only letters, number, or dashes',
            default: 'admin',
            required: true
        },
        APP_USE_PROXY: {
            description: 'Please select proxy system',
            type: 'string',
            message: 'Supported proxy system: nginx',
            default: 'none',
            required: false
        },
        APP_COOKIE_NAME: {
            description: 'Please enter app cookie name',
            type: 'string',
            message: '',
            default: `laco.${helper.randomstring.generate(10)}.sid`,
            required: false
        },
        DB_CONNECT_URI: {
            description: 'Enter mongodb connect uri',
            type: 'string',
            default: 'mongodb://laco_user:123456789@localhost/laco',
            required: true
        },
        DB_OPTIONS_poolSize: {
            description: 'Enter mongodb poolsize',
            type: 'integer',
            default: 10,
            required: true
        },
        ASSETS_BUNDLING: {
            description: 'Enable assets bundling',
            //type: 'integer',
            pattern: /^(0|1)$/,
            message: '0: Disable | 1: Enable',
            default: 0,
            required: true
        },
        ASSETS_MINIFY: {
            description: 'Enable assets minify',
            //type: 'integer',
            pattern: /^(0|1)$/,
            message: '0: Disable | 1: Enable',
            default: 0,
            required: true
        },
        ASSETS_FINGERPRINT: {
            description: 'Enable assets finger print',
            //type: 'integer',
            pattern: /^(0|1)$/,
            message: '0: Disable | 1: Enable',
            default: 0,
            required: true
        }
    }
};

exports.rules = rules;
