// laco-ext-router/index.js
"use strict";
const path = require('path');
const eventEmitter = lacoLib('./libs/eventEmitter');
const router = lacoLib('./libs/router');
const logger = lacoLib('./libs/logger');
const adminRouter = require('./admin');
const Extension = lacoLib('./libs/extension');
const loadedExts = lacoLib('./libs/ext_loader').loadedExts;
const user = lacoLib('./libs/user');

let handleHomeGet = (req, res) => {
    res.send(res.abc + ' home handler outside');
}

let _laco = null;

class ExtRouter extends Extension {
    constructor(laco) {
        super(__dirname, 'Router ext');
        _laco = laco;
        this.registerViews([
            'login',
            'register'
        ]);

        const loginAsset = path.join(LACO_ROOT, 'data/assets/admin');
        const depPaths = [
            path.join(loginAsset, 'plugins/material-design-iconic-font/css/material-design-iconic-font.min.css'),
            path.join(loginAsset, 'plugins/bootstrap/css/bootstrap.css'),
            path.join(loginAsset, 'plugins/node-waves/waves.css'),
            path.join(loginAsset, 'plugins/animate-css/animate.css'),
            path.join(loginAsset, 'css/style.css'),
            path.join(loginAsset, 'css/custom.css'),
            path.join(loginAsset, 'plugins/jquery/jquery.min.js'),
            path.join(loginAsset, 'plugins/bootstrap/js/bootstrap.js'),
            path.join(loginAsset, 'plugins/node-waves/waves.js'),
            path.join(loginAsset, 'plugins/jquery-validation/jquery.validate.js'),            
            path.join(loginAsset, 'js/admin.js'),
            path.join(loginAsset, 'js//pages/examples/sign-in.js')
        ];
        let loginDeps = [];
        for(let dep of depPaths) {
            loginDeps.push({
                path: dep,
                'if-flag': 'laco-login'
            });
        }
        this.addDependencies(loginDeps);
        this.addPageFlags('laco-login');

        adminRouter.init(laco);
        laco.app.use('/' + laco.nconf.get('APP_ADMIN_DIR'), adminRouter.middleware());

        router.register({
            name: 'home',
            label: 'Home page',
            priority: 2,
            path: '/',
            get: handleHomeGet,
        });
        router.register({
            name: 'login',
            label: 'Login',
            priority: 3,
            path: '/login',
            get: (req, res, next) => loadedExts.router.login(req, res, next),
            post: (req, res, next) => this.login(req, res, next)
        });
        router.register({
            name: 'logout',
            label: 'Logout',
            priority: 3,
            path: '/logout',
            get: (req, res, next) => {
                req.session.user = null;
                req.flash('success', `You have already logged out!`);
                return res.redirect('/login');
            }
        });
    }
    
    login(req, res) {
        let data = {};
        if(req.method == 'GET') return this.view('login', {}, req, res);
        if(req.method == 'POST') {
            user.check(req.body.username, req.body.password).then(_user => {
                req.session.user = _user;
                res.redirect(req.query.back || _laco.nconf.get('APP_ADMIN_DIR'));
            }).catch(err => {
                req.flash('danger', err);
                res.redirect(req.originalUrl);
            });
        }
        else res.end('Invalid action!');
    }
}

module.exports = ExtRouter;
